tpke/src/ciphertext.rs

```diff
@@ -32,17 +32,47 @@ pub struct Ciphertext<E: Pairing> {
 }
 
 impl<E: Pairing> Ciphertext<E> {
+    pub fn check(&self, aad: &[u8], g_inv: &E::G1Prepared) -> Result<bool> {
+        self.header()?.check(aad, g_inv)
+    }
+
+    pub fn ciphertext_hash(&self) -> [u8; 32] {
+        sha256(&self.ciphertext)
+    }
+
+    pub fn header(&self) -> Result<CiphertextHeader<E>> {
+        Ok(CiphertextHeader {
+            commitment: self.commitment,
+            auth_tag: self.auth_tag,
+            ciphertext_hash: self.ciphertext_hash(),
+        })
+    }
+    pub fn payload(&self) -> Vec<u8> {
+        self.ciphertext.clone()
```

Comment:
Why clone?

Reply:
Because `&self.ciphertext` is hidden behind `&self` reference, and we need to take the ownership of those bytes to be used in bindings